plugin/role_set_utils.go

```diff
@@ -0,0 +1,161 @@
+		}
+	}
+
+	if boundResources.tokenGen != nil {
+		return b.addServiceAccountKeyWal(ctx, req, boundResources.rolesetName, boundResources.accountId, boundResources.tokenGen.KeyName)
+	}
+	return nil
+}
+
+// addServiceAccountKeyWal creates WAL to clean up a roleset service account key (for access tokens) if needed.
+func (b *backend) addServiceAccountKeyWal(ctx context.Context, req *logical.Request, rolesetName string, accountId *gcputil.ServiceAccountId, keyName string) error {
+	if accountId == nil {
+		b.Logger().Debug("skip adding service account key WAL for nil account")
+		return nil
+	}
+
+	_, err := framework.PutWAL(ctx, req.Storage, walTypeAccount, &walAccountKey{
+		RoleSet:            rolesetName,
+		ServiceAccountName: accountId.ResourceName(),
+		KeyName:            keyName,
+	})
+	if err != nil {
+		return errwrap.Wrapf("unable to create WAL entry to clean up service account key: {{err}}", err)
+	}
+	return nil
+}
+
+// tryDeleteRoleSetResources tries to delete GCP resources previously managed by a roleset.
+// This assumes that deletion of these resources will already be guaranteed by WALs and will return errors
+// as a list of warnings instead.
+func (b *backend) tryDeleteRoleSetResources(ctx context.Context, req *logical.Request, boundResources *roleSetResources) []string {
+	if boundResources == nil {
+		b.Logger().Debug("skip deletion for nil roleset resources")
+		return nil
+	}
+
+	httpC, err := b.HTTPClient(req.Storage)
+	if err != nil {
+		return []string{err.Error()}
+	}
+
+	iamAdmin, err := iam.NewService(ctx, option.WithHTTPClient(httpC))
+	if err != nil {
+		return []string{err.Error()}
+	}
+
+	iamHandle := iamutil.GetIamHandle(httpC, useragent.String())
+
+	warnings := make([]string, 0)
+	if boundResources.accountId != nil {
+		if err := b.deleteTokenGenKey(ctx, iamAdmin, boundResources.tokenGen); err != nil {
+			w := fmt.Sprintf("unable to delete key under service account %q (WAL entry to clean-up later has been added): %v", boundResources.accountId.ResourceName(), err)
+			warnings = append(warnings, w)
+		}
+
+		if err := b.deleteServiceAccount(ctx, iamAdmin, boundResources.accountId); err != nil {
+			w := fmt.Sprintf("unable to delete service account %q (WAL entry to clean-up later has been added): %v", boundResources.accountId.ResourceName(), err)
+			warnings = append(warnings, w)
+		}
+
+		if merr := b.removeBindings(ctx, iamHandle, boundResources.accountId.EmailOrId, boundResources.bindings); merr != nil {
+			for _, err := range merr.Errors {
+				w := fmt.Sprintf("unable to delete IAM policy bindings for service account %q (WAL entry to clean-up later has been added): %v", boundResources.accountId.EmailOrId, err)
+				warnings = append(warnings, w)
+			}
+		}
+	}
+
+	return nil
+}
+
+func emailForServiceAccountName(project, accountName string) string {
+	return fmt.Sprintf(serviceAccountEmailTemplate, accountName, project)
+}
+
+func generateAccountNameForRoleSet(rsName string) (name string) {
```

Comment:
I realize it's private, but can you document this with the expected input and expected output? There's a lot of string magic so have 1-2 examples of "here's what this function will output" could be helpful